File: runtime/app/policy.py
from app.models import EvaluateRequest, EvaluateResponse, PolicyDecision, RiskTier


def evaluate_policy(request: EvaluateRequest) -> EvaluateResponse:
    max_actions_per_hour = int(request.policy.get("maxActionsPerHour", 20))
    max_linkedin_messages = int(request.policy.get("maxLinkedinMessages", 30))
    require_approval_tier = int(request.policy.get("requireApprovalTier", 3))

    if request.expected_actions_this_hour > max_actions_per_hour:
        return EvaluateResponse(
            decision=PolicyDecision.DENY,
            reason="Exceeded global actions-per-hour policy cap.",
        )

    if "linkedin" in request.action and request.expected_linkedin_messages > max_linkedin_messages:
        return EvaluateResponse(
            decision=PolicyDecision.DENY,
            reason="Exceeded LinkedIn message policy cap.",
        )

    risk_number = int(request.risk_tier.split("_")[1])
    if risk_number >= require_approval_tier:
        return EvaluateResponse(
            decision=PolicyDecision.REQUIRE_APPROVAL,
            reason=f"Risk tier {request.risk_tier} requires manual approval.",
        )

    return EvaluateResponse(
        decision=PolicyDecision.ALLOW,
        reason="Action is within autonomous policy envelope.",
    )


def infer_risk_tier(action: str) -> RiskTier:
    if "linkedin" in action and ("message" in action or "connect" in action):
        return RiskTier.TIER_3

    if "extract" in action:
        return RiskTier.TIER_1

    return RiskTier.TIER_2

File: web/src/lib/policy.ts
import type { AgentKind, Policy } from "@prisma/client";
import type { PolicyDecision, RiskTier } from "@/lib/contracts";

export function mapRiskTier(kind: AgentKind, action: string): RiskTier {
  if (kind === "linkedin") {
    if (action.includes("message") || action.includes("connect")) {
      return "tier_3";
    }
    return "tier_2";
  }

  if (kind === "browser") {
    return action.includes("extract") ? "tier_1" : "tier_2";
  }

  return "tier_2";
}

export function evaluatePolicy(
  policy: Pick<Policy, "maxActionsPerHour" | "maxLinkedinMessages" | "requireApprovalTier">,
  args: { kind: AgentKind; action: string; expectedActionsThisHour: number },
): PolicyDecision {
  const riskTier = mapRiskTier(args.kind, args.action);
  const tierNumber = Number(riskTier.split("_")[1]);

  if (args.expectedActionsThisHour > policy.maxActionsPerHour) {
    return "deny";
  }

  if (args.kind === "linkedin" && args.expectedActionsThisHour > policy.maxLinkedinMessages) {
    return "deny";
  }

  if (tierNumber >= policy.requireApprovalTier) {
    return "require_approval";
  }

  return "allow";
}

